Add support for primary constructors in LoggerMessageGenerator

kimsey0 opened the pull request:
Fixes #91121.

kimsey0 committed:
Add support for primary constructors in LoggerMessageGenerator

kimsey0 commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -71,527 +71,528 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
+                Dictionary<SyntaxTree, SemanticModel> semanticModels = classes
+                    .Select(c => c.SyntaxTree)
+                    .Distinct()
+                    .ToDictionary(st => st, st => _compilation.GetSemanticModel(st));
```
This assumes that the `ClassDeclarationSyntax.SyntaxTree` returns a reference to the same instance for class declarations in the same syntax tree.

kimsey0 commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -624,14 +625,55 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
                 bool onMostDerivedType = true;
 
+#if ROSLYN4_8_OR_GREATER
+                IEnumerable<ClassDeclarationSyntax> primaryConstructors = classType.InstanceConstructors
```
Unfortunately, there seems to be no way to identify primary constructor parameters in the semantic model, and the partial class declaration that contains the primary constructor may not contain a method annotated with `[LoggerMessage]`, thereby not being picked up by the `SyntaxProvider` in `LoggerMessageGenerator.Roslyn4.0.cs`, so we have to work backwards from here, potentially parsing and creating semantic models for class declarations that weren't otherwise visited.

tarekgh replied:
@CyrusNajmabadi @eiriktsarpalis any suggestion about the best way to handle that?

kimsey0 replied:
@333fred mentioned in the Discord that this could be potentially be done with `LookupSymbols` or a speculative model, but:
1. `LookupSymbols` doesn't return primary constructor parameters for any positions I know will exist in all cases, even in minimal class declarations with just one partial method declaration.
2. `TryGetSpeculativeSemanticModelForMethodBody` requires a position parameter _within an existing method body_ which, once again, may not exist. https://github.com/dotnet/roslyn/issues/61564 suggests speculative models don't work in this case.

kimsey0 commented:
The GitHub diff tool seems to have picked up a lot of nonexistent changes in `LoggerMessageGenerator.Parser.cs` which are caused by removing a level of indentation. I don't know how to fix that. The only actual changes are around lines 74-78 and in `FindLoggerField`.

kimsey0 commented:
@333fred mentioned in the discord that this can probably be done better with the `LookupSymbols` API. I'll have a look and edit the pull request accordingly.

tarekgh commented:
@kimsey0 thanks for submitting the PR. I converted it to draft till it is ready for actual review.

CyrusNajmabadi commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -624,14 +625,55 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
+
+                        foreach (ParameterSyntax parameter in parameters)
+                        {
+                            if (semanticModel.GetSymbolInfo(parameter.Type).Symbol is not ITypeSymbol parameterType)
```
you don't need to go to the semantic model for this.  the INmaedTypeSymbol for the class can get you the IMethodSymbol for the primary cosntructor

which has the IParameterSyjmbols.

kimsey0 replied:
Thanks. I'll have a look.

kimsey0 replied:
That worked nicely. I've changed it now. Thank you.

kimsey0 committed:
Get the primary constructor parameters types from the constructor symbol instead of from the semantic model

kimsey0 commented:
@tarekgh, with the helpful suggestion from Cyrus, I now think this is ready for review. Should I mark it as such or wait for more feedback, perhaps from @eiriktsarpalis? (I don't know when he's back from vacation.)

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -632,6 +632,30 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
 
                 bool onMostDerivedType = true;
 
+#if ROSLYN4_8_OR_GREATER
```
>ROSLYN4_8_OR_GREATER [](http://example.com/codeflow?start=4&length=20)

I am seeing we are defining such constants in the csproj [https://github.com/dotnet/runtime/blob/017593d90781d797df8b2241f6d1f83c236c442b/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/Microsoft.Extensions.Logging.Generators.Roslyn4.4.csproj#L6](https://github.com/dotnet/runtime/blob/017593d90781d797df8b2241f6d1f83c236c442b/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/Microsoft.Extensions.Logging.Generators.Roslyn4.4.csproj#L6)

Could you try to use `ROSLYN4_4_OR_GREATER` instead and look if this will work? I expect we use recent Roslyn which I expect it will work.

kimsey0 replied:
That didn't work with the previous version of this pull request, since the `ParameterList` property on `ClassDeclarationSyntax` isn't available until Roslyn 4.8, but it may be possible now that this uses the constructor parameters on the class symbol instead. Let me give it a try!

tarekgh replied:
Look at my suggestion https://github.com/dotnet/runtime/pull/101660#discussion_r1585301628. I am asking if we can do that instead of creating a new version of the library just to match the desired Roslyn version.

333fred commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -632,6 +632,30 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
+                    .Where(ic => ic.DeclaringSyntaxReferences
+                        .Any(ds => ds.GetSyntax() is ClassDeclarationSyntax));
+
+                foreach (IMethodSymbol primaryConstructor in primaryConstructors)
```
You should also test scenarios like this:

```cs
class Test(ILogger logger)
{
    private readonly ILogger _logger = logger;
}
```

and ensure that you're using the right logger for that scenario.

tarekgh replied:
@333fred is it possible to lookup the fields first and then if didn't find any try to look up the primary constructors? 

Also, for detecting the primary constructor, should check `constructor.Body == null && constructor.ExpressionBody == null`?

tarekgh replied:
> private readonly ILogger _logger = logger;

good test as we don't support private fields.

333fred replied:
> @333fred is it possible to lookup the fields first and then if didn't find any try to look up the primary constructors?

Sure, that would be possible. Talking with Aleksey, that's likely the approach you want to take in general. However, I'm going to strongly recommend that you explicitly consider these cases and lay out, in comments and documentation, the exact rules you follow. IE, what happens when a primary constructor parameter is shadowed? What happens when a base type has a protected member that is named differently than the parameter (ie, has an `_`)? This will make it easier to verify the behavior is what you're expecting, and to handle bugs as by-design or not.

> Also, for detecting the primary constructor, should check `constructor.Body == null && constructor.ExpressionBody == null`?

You only need the check that kimsey0 put in already. It's important to note that you can only detect primary constructors from source, you cannot detect them from metadata.

> good test as we don't support private fields.

Wait, what? I'll reiterate that I think you need spell out exactly what the rules are, in comments above the code, so that you can verify the code follows them. I thought I might understand the rules, but clearly I do not.

kimsey0 replied:
@333fred, that sounds excellent. I'd be very happy to use a new `LookupSymbols` API for this. The pull request isn't exactly time sensitive either. I'd just be happy for it to land in .NET 9.

For now, I'll add the test cases you've suggested and try to find ways to handle them, if nothing else, then for the sake of my own learning.

tarekgh replied:
> Wait, what? I'll reiterate that I think you need spell out exactly what the rules are, in comments above the code, so that you can verify the code follows them. I thought I might understand the rules, but clearly I do not.

I am wrong here, looking more seeing we can support private fields https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator#basic-usage this is because we generate the code inside the same class and LoggerMessage is not static. Sorry for confusion.

tarekgh replied:
> Should that discussion happen in this pull request or perhaps in https://github.com/dotnet/runtime/issues/91121?

Either way is fine. we can continue discussing in the PR and then put the conclusion in the issue. I am fine to move the discussion to the issue too. 

> Your comments above expose some edge cases, but there are probably more that I may not think of.

I think when applying the suggestion which looking up the fields first will address most these corner cases. Of course we'll need to document the behavior.

kimsey0 replied:
I have added the three test cases suggested by @333fred, as well as one that shadows the primary constructor parameter with an `ILogger` field:

```
class Test(ILogger logger)
{
    private readonly ILogger logger = logger;
}
```

I've changed the code to support these cases by:
1. First looking for an `ILogger` field.
2. If none is found, look for an `ILogger` primary constructor parameter.
3. Ignore any primary constructor parameters which are shadowed by a field.

In practice, this means that:
1. Fields are prioritized over primary constructor parameters.
2. Shadowed primary constructor parameters are treated as if they don't exist.

I think this is a reasonable behavior, though one could argue that it would make sense to have a specific warning when a `ILogger` primary constructor parameter is found that is shadowed by a non-`ILogger` field.

How should this behavior be documented? Just on [the learn page](https://github.com/dotnet/docs/blob/main/docs/core/extensions/logger-message-generator.md)?

tarekgh replied:
> though one could argue that it would make sense to have a specific warning when a ILogger primary constructor parameter is found that is shadowed by a non-ILogger field.

Yes, we can have a diagnostic for that. Would it be better to have it as `Info` instead of `warning`? 

> How should this behavior be documented? Just on [the learn page](https://github.com/dotnet/docs/blob/main/docs/core/extensions/logger-message-generator.md)?

Right, this is the right place to document the behavior. We'll need to update this doc anyway for primary constructor usage with a sample.

kimsey0 replied:
> Yes, we can have a diagnostic for that. Would it be better to have it as Info instead of warning?

I think that would make sense if we then still report the _no logger field found_ error in this case. I'll try implementing that.

> Right, this is the right place to document the behavior. We'll need to update this doc anyway for primary constructor usage with a sample.

Should I create a pull request over there before this merges, or is it fine to do after?

tarekgh replied:
It is fine we can add the doc after merging this PR.

kimsey0 replied:
Sounds good. If both the behavior described above as well as the code implementing it is acceptable to all parties, I think this is ready for review again.

kimsey0 replied:
I forgot to commit and push the info diagnostic, but just did that. I don't know if there are any guidelines for how exactly these diagnostic messages should be written or how the translation flow goes? (We can also just revert this. I think it's useful, but not necessary.)

tarekgh replied:
You don't have to worry about the translation of the added messages. I am seeing the added diagnostic is useful to catch the concerned case.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/Microsoft.Extensions.Logging.Generators.Roslyn4.8.csproj:
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
@ericstj do you see a problem if we upgrade the 

```Xnl
<RoslynApiVersion>$(MicrosoftCodeAnalysisVersion_4_4)</RoslynApiVersion>

ericstj replied:
If we update to that Roslyn version it means we remove the one that targets 4.4.

Updating for the inbox generators in 9.0 could be OK (if we can guarantee that 4.8 is present in all SDK/VS versions where 9.0 is supported).

Updating for the nuget package generators is ok so long as we don't take away any features / perf from folks using roslyn 4.4.  Of course that's moot if noone is using 4.4-4.7.

Somewhere there is a matrix of compiler versions that are supported that could help answer this question.

kimsey0 replied:
Looks like it doesn't matter in the end. I don't actually need to have code in the generator that's conditionally compiled for Roslyn 4.8. It's only in the test project I need to be sure primary constructors are supported. @tarekgh, is this approach of just upgrading the `RoslynApiVersion` in `Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj` OK?

tarekgh replied:
Can you add a new `Microsoft.Extensions.Logging.Generators.Roslyn4.8.Tests.csproj` that target Roslyn 4.8?

kimsey0 replied:
I had one, but removed it in the last commit since you asked me to revert all changes to the solution file, where I added it. Shouldn't I add it to the solution file if I add a new test project?

tarekgh replied:
Yes, if adding a new test project, it will be fine to add it to the solution.

kimsey0 committed:
Prioritize fields over primary constructor parameters and ignore shadowed parameters when finding a logger

kimsey0 commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln:
```diff
@@ -15,6 +15,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Extensions.Loggin
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Extensions.Logging.Generators.Roslyn4.8", "gen\Microsoft.Extensions.Logging.Generators.Roslyn4.8.csproj", "{40FA38ED-D2C9-4608-B85E-F731655C1864}"
+EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Extensions.Logging.Abstractions", "ref\Microsoft.Extensions.Logging.Abstractions.csproj", "{7F536552-0E2A-4642-B7CF-863727C2F9CD}"
```
Shouldn't I keep a reference to the new test project?

kimsey0 committed:
Make checking for primary constructors non-conditional on Roslyn version and simplify project setup

tarekgh commented:
@geeknoid @joperezr to be aware about this change if need to update the extensions source logging too.

joperezr commented:
Thanks for flagging this @tarekgh! It does seem like it is something we'll want to do in extensions. Logged https://github.com/dotnet/extensions/issues/5178 to track

kimsey0 committed:
Reintroduce Roslyn 4.8 test project

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -651,10 +654,48 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
+                    {
+                        if (IsBaseOrIdentity(parameter.Type, loggerSymbol))
+                        {
+                            if (shadowedNames.Contains(parameter.Name))
```
>shadowedNames [](http://example.com/codeflow?start=32&length=13)

I am not sure if shadowedNames is needed at all? If we encountered any field of type `loggerSymbol`, we'll not reach here at all because we'll return `return (loggerField, false);`.

kimsey0 replied:
> If we encountered any field of type `loggerSymbol`, we'll not reach here at all because we'll return (loggerField, false);.

Yes, that is correct. This is for the case where we do not encounter any fields of type `loggerSymbol`, but where one of the non-logger fields shadows a logger primary constructor parameter. Here's a relevant test case:

```csharp
partial class C(ILogger logger)
{
    private readonly object logger = logger;

    [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = ""M1"")]
    public partial void M1();
}
```

In this case, we can't use the `logger` primary constructor parameter, since there's no way to refer to it.

kimsey0 committed:
Add info-level diagnostic for logger primary constructor parameters that are shadowed by field

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/DiagnosticDescriptors.cs:
```diff
@@ -208,5 +208,13 @@ public static class DiagnosticDescriptors
             defaultSeverity: DiagnosticSeverity.Error,
             isEnabledByDefault: true);
+
+        public static DiagnosticDescriptor PrimaryConstructorParameterLoggerHidden { get; } = DiagnosticDescriptorHelper.Create(
```
Please add the entry to the file [https://github.com/dotnet/runtime/blob/0588f245dba6418349c22894b40337afc34f1cfc/docs/project/list-of-diagnostics.md?plain=1#L148](https://github.com/dotnet/runtime/blob/0588f245dba6418349c22894b40337afc34f1cfc/docs/project/list-of-diagnostics.md?plain=1#L148) too.

tarekgh replied:
and please tracking updating the doc [https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/source-generator-overview#reference](https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/source-generator-overview#reference) too.

kimsey0 replied:
I created https://github.com/dotnet/docs/issues/41084 to track the necessary documentation updates.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -651,10 +654,50 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
                             }
                         }
+
+                        shadowedNames.Add(fs.Name);
```
>shadowedNames.Add(fs.Name); [](http://example.com/codeflow?start=24&length=27)

should have this with `else`? we should carry the shadowed names that not are loggerSymbol.

kimsey0 replied:
It doesn't make any difference for the behavior, since when a logger field is found, the primary constructor code isn't run, but you're right that we may as well avoid the allocation.

tarekgh commented:
>             Assert.Equal("SYSLIB1026", diagnostics[0].Id);

Please add similar test for the new diagnostic we added. 

---
Refers to: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorParserTests.cs:1115 in 6cf2b2f. [](commit_id = 6cf2b2f33f281f1fd82620f71bc0edc8bd317baf, deletion_comment = False)

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -651,10 +654,50 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
                             }
                         }
+
+                        shadowedNames.Add(fs.Name);
```
>shadowedNames [](http://example.com/codeflow?start=24&length=13)

Could you please add a comment explaining what this shadowedNames for? You may put your example you mentioned before in the comment. Thanks!

kimsey0 replied:
Sure thing!

tarekgh commented:
@kimsey0 I added a few minor comments. I think we should be good to go after addressing these. Thanks!

kimsey0 committed:
Update list of diagnostics with new logging message generator diagnostic

kimsey0 committed:
Only add non-logger field names to set of shadowed names

kimsey0 committed:
Add comment explaining the use of the set of shadowed names with an example

kimsey0 commented:
> > ```
> >         Assert.Equal("SYSLIB1026", diagnostics[0].Id);
> > ```
> 
> Please add similar test for the new diagnostic we added.

I added assertions that the new diagnostic is reported to [the existing test cases with shadowed primary constructor parameters](https://github.com/dotnet/runtime/pull/101660/files#diff-9a439d24d74b4541df08370e1f988d75640836b9b9e07009d96be4b019706406R507-R553). Would you like any additional tests of this?

tarekgh reviewed (approved):
Thanks @kimsey0 providing the fix!
